src/mesh/forest/tree.cpp

```diff
@@ -0,0 +1,289 @@
+//========================================================================================
+// (C) (or copyright) 2024. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+
+#include <algorithm>
+#include <array>
+#include <limits>
+#include <map>
+#include <memory>
+#include <set>
+#include <stack>
+#include <tuple>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+#include "basic_types.hpp"
+#include "defs.hpp"
+#include "mesh/forest/relative_orientation.hpp"
+#include "mesh/forest/tree.hpp"
+#include "mesh/logical_location.hpp"
+#include "utils/bit_hacks.hpp"
+#include "utils/indexer.hpp"
+
+namespace parthenon {
+namespace forest {
+
+Tree::Tree(Tree::private_t, std::int64_t id, int ndim, int root_level, RegionSize domain,
```

Comment:
What's going on with the `Tree::private_t` first argument? I don't think I've seen that before.

Reply:
This enforces that `Tree`s can only be created as `shared_ptr`s via `Tree::create`. If the constructor is just made `private` or `protected`, then the call to `std::make_shared<Tree>` in `Tree::create` fails because the constructor is not accessible. This pattern makes the constructor effectively private while still allowing the use of `make_shared`.

Reply:
Ah ok. Thanks for the explanation!

Reply:
Might be worth adding this explanation to the code as comment (also to educate other devs like me about this pattern when reading the code again at a later point ;) )